lib/mlir/Dialect/Clift/IR/ImportModel.cpp

```diff
@@ -0,0 +1,599 @@
+    if (not UnderlyingType)
+      rc_return nullptr;
+    rc_return make<clift::TypedefAttr>(ModelType.ID(),
+                                       ModelType.OriginalName(),
+                                       UnderlyingType);
+  }
+
+  RecursiveCoroutine<clift::TypeDefinition>
+  getTypeAttribute(const model::UnionType &ModelType,
+                   const bool RequireComplete) {
+    if (not RequireComplete) {
+      const auto T = clift::UnionType::get(Context,
+                                           ModelType.ID(),
+                                           ModelType.OriginalName());
+      if (not T.isDefinition())
+        IncompleteTypes.try_emplace(ModelType.ID(), &ModelType);
+      rc_return T;
+    }
+
+    DefinitionGuard Guard(*this, ModelType.ID());
+    if (not Guard) {
+      if (EmitError)
+        EmitError() << "Recursive definition of UnionType " << ModelType.ID();
+      rc_return nullptr;
+    }
+
+    AttributeVector<clift::FieldAttr> Fields;
+    Fields.reserve(ModelType.Fields().size());
+
+    for (const model::UnionField &Field : ModelType.Fields()) {
+      auto const FieldType = rc_recur
+        getQualifiedType(Field.Type(), /*RequireComplete=*/true);
+      if (not FieldType)
+        rc_return nullptr;
+      auto const Attribute = make<clift::FieldAttr>(0,
+                                                    FieldType,
+                                                    Field.CustomName());
+      if (not Attribute)
+        rc_return nullptr;
+      Fields.push_back(Attribute);
+    }
+
+    rc_return make<clift::UnionType>(ModelType.ID(),
+                                     ModelType.OriginalName(),
+                                     Fields);
+  }
+
+  RecursiveCoroutine<clift::TypeDefinition>
+  getTypeAttribute(const model::Type &ModelType, bool &RequireComplete) {
+    switch (ModelType.Kind()) {
+    case model::TypeKind::CABIFunctionType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::CABIFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::EnumType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::EnumType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::RawFunctionType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::RawFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::StructType: {
+      auto const &T = llvm::cast<model::StructType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::TypedefType: {
+      auto const &T = llvm::cast<model::TypedefType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::UnionType: {
+      auto const &T = llvm::cast<model::UnionType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    default:
+      revng_abort();
+    }
+  }
```

Comment:
The function returns a `RecursiveCoroutine`, so all the `return`s should be `rc_return rc_recur getTypeAttribute`.

I'm honestly surprised this even compiles. My expectation was that the compiler should warn you or plainly error out :thinking:

Reply:
`RecursiveCoroutine` is just a regular object and returning one from a function is just fine. In a coroutine this object is constructed implicitly by the compiler. We're just returning the object returned by another coroutine. Think of this as a coroutine tail call. It's just more efficient this way.

Reply:
Fair enough. It's just not something I had in mind when I designed them, but it checks out!

Finally someone is using them besides me, and this brings up new ways to use them that I did not envision :)